### PR TITLE
only allow primitive values in context keys

### PR DESCRIPTION
**Pull request by jrieken**
Alternative approach to https://github.com/microsoft/vscode/pull/147734, tho doesn't compile yet because folks have been creative...

### CODE REVIEW

**Comment by alexr00** [comment]
For both of the `tunnelView.ts` errors I do use several properties on the object and for one of them I do object comparison. It all looks removable though, so I can change it if needed.

**Comment by bpasero** [comment]
@sandy081 for history search widget via https://github.com/microsoft/vscode/blob/21b0b98264376f9844acecf60a23f808f0fe7428/src/vs/platform/widget/common/contextScopedWidget.ts

**Comment by joaomoreno** [comment]
Finally you caught me. 🙈 I might even have been the one pointing @alexr00 and @sandy081 to doing that.

In SCM we actually put live objects in there. It makes things really easy [when implementing commands](https://github.com/microsoft/vscode/blob/93c682fe068bc5517e2a04b7ef9847f39d917f30/src/vs/workbench/contrib/scm/browser/scm.contribution.ts#L314-L316). We can change it to use a string identifier and just have the commands get the repository from the service using that identifier.

**Comment by joaomoreno** [comment]
SCM: https://github.com/microsoft/vscode/pull/148047

**Comment by sandy081** [comment]
history widget needs list service/widget similar implementation to get last focused history widget and run the commands. Will look into it next milestone.

**Comment by joaomoreno** [comment]
@sandy081 I was never a fan of the list "last focused" strategy. A better approach is to follow what I just did for SCM: have the context key value be a string id, which you could come to some service and say "get me the history widget with this id".

**Comment by jrieken** [comment]
> have the context key value be a string id, which you could come to some service and say "get me the history widget with this id".

I don't know how I feel about this. The primary purpose of context keys is to collect value to evaluate against "written conditions", e.g mainly for menu items and keybindings. Using them as a "wormwhole" to transport information across the system seem fishy to me

**Comment by joaomoreno** [comment]
> I don't know how I feel about this. The primary purpose of context keys is to collect values to evaluate against "written conditions", e.g mainly for menu items and keybindings. Using them as a "wormwhole" to transport information across the system seem fishy to me

Going back/forward on the history of an input needs to run as a command so people can override keyboard shortcuts. A keyboard shortcut runs the command under the context of the currently focused DOM node. There's no other additional possible context that we can have. So, from the context key service we must go to the "live object" which handles history for that specific input box.

The support we lack here is ability to couple configurable keyboard shortcuts with live objects. All we have right now is the context key service.

**Comment by jrieken** [comment]
> So, from the context key service we must go to the "live object" which handles history for that specific input box. I don't know any other way to implement this.

Isn't this the same as how list/tree commands discover what instance to operate on? Or how editor commands know what editor instance to operate on? They all have some kind of service in the back that know the active list, tree, or editor.

**Comment by joaomoreno** [comment]
Yeah. List/tree commands do what @sandy081 mentioned: there's a global service tracking which list/tree was last focused and picks that on deman. I've always found that to be weird to reason about. I also never really learned to trust a focus tracking mechanism. Conceptually, it makes much more sense to me to piggy back on the tree nature of the DOM for contextual information, which is what the context key service gives us. By setting the right context on the right UI areas, we can have multiple simultaneous different contexts without active focus tracking code. When the event happens, the right context is found.

I feel it weird that an event happens deep down in the DOM, in a place rich in context, but control flow bubbles all the way up to a global service which somehow assumes it knows which live object to pick. Why wouldn't control flow simply bubble to the point that has the context and uses it directly? That's what we do today by keeping live SCMRepository instances as values in scoped context key services.

**Comment by jrieken** [comment]
> I also never really learned to trust a focus tracking mechanism 

You should learn to trust it: every time you press anything like arrow, backspace, enter, tab etc in the editor you are using that mechanism. Also, it doesn't work by tracking, it simply knows all widgets and checks which of them [has focus](https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/editor/browser/editorExtensions.ts#L255).

**Comment by jrieken** [comment]
@joaomoreno @alexr00 @sandy081 Please unassign yourself when we are done here. I would like to get this done during this debt week. Thanks

**Comment by sandy081** [comment]
✅  from my end

**Comment by alexr00** [comment]
Forgot to un-assign myself. Mine should be good too.

**Comment by jrieken** [comment]
Thanks everyone!

File: src/vscode-dts/vscode.d.ts [hunk 1]
```diff
@@ -15680,7 +15680,7 @@ declare module 'vscode' {
 		readonly viewColumn: ViewColumn;
 
 		/**
-		 * The active {@link Tab tab} in the group. This is the tab whose contents are currently
+		 * The active {@link Tab tab} in the group. This is the tab which contents are currently
```
Comment (alexdima) [inline src/vscode-dts/vscode.d.ts:15683]: Was this an intentional change for this PR?

**Review by alexdima — APPROVED** [review]
Thank you for tackling this!

**Pull request merged**